state-chain/pallets/cf-lending-pools/src/general_lending.rs

```diff
@@ -0,0 +1,1635 @@
+					else {
+						// Don't swap yet if we can't determine asset's price
+						continue;
+					};
+
+					if fee_usd_value >= config.fee_swap_threshold_usd {
+						let fees_to_swap = core::mem::take(fee_amount);
+						let swap_request_id = initiate_swap::<T>(
+							*collateral_asset,
+							fees_to_swap,
+							pool_asset,
+							LendingSwapType::FeeSwap { pool_asset },
+							config.fee_swap_max_oracle_slippage,
+						);
+
+						Pallet::<T>::deposit_event(Event::LendingPoolFeeCollectionInitiated {
+							asset: pool_asset,
+							swap_request_id,
+						});
+					}
+				}
+			});
+		}
+
+		// Additionally swapp all network fee contributions from fees:
+		for asset in PendingNetworkFees::<T>::iter_keys().collect::<Vec<_>>() {
+			PendingNetworkFees::<T>::mutate(asset, |fee_amount| {
+				// NOTE: if asset is FLIP, we shouldn't need to swap, but it should still work,
+				// and it seems easiest to not write a special case (esp if we only support
+				// boost for BTC)
+				if *fee_amount > 0 {
+					let swap_request_id =
+						T::SwapRequestHandler::init_network_fee_swap_request(asset, *fee_amount);
+
+					Pallet::<T>::deposit_event(Event::LendingNetworkFeeCollectionInitiated {
+						swap_request_id,
+					});
+				}
+				*fee_amount = 0;
+			});
+		}
+	}
+
+	Weight::zero()
+}
+
+impl<T: Config> LendingApi for Pallet<T> {
+	type AccountId = T::AccountId;
+
+	/// Create a new loan (assigning a new loan id) provided that the account's existing collateral
+	/// plus any `extra_collateral` is sufficient. Will update the primary collateral asset if
+	/// provided.
+	#[transactional]
+	fn new_loan(
+		borrower_id: T::AccountId,
+		asset: Asset,
+		amount_to_borrow: AssetAmount,
+		primary_collateral_asset: Option<Asset>,
+		extra_collateral: BTreeMap<Asset, AssetAmount>,
+	) -> Result<LoanId, DispatchError> {
+		ensure!(
+			T::SafeMode::get().borrowing_enabled.contains(&asset),
+			Error::<T>::LoanCreationDisabled
+		);
+
+		let config = LendingConfig::<T>::get();
+
+		let loan_id = NextLoanId::<T>::get();
+		NextLoanId::<T>::set(loan_id + 1);
+
+		LoanAccounts::<T>::mutate(&borrower_id, |maybe_account| {
+			let account = Self::create_loan_account_if_empty(
+				borrower_id.clone(),
+				maybe_account,
+				primary_collateral_asset,
+			)?;
+
+			let primary_collateral_asset = account.primary_collateral_asset;
+
+			ensure!(
+				extra_collateral.contains_key(&primary_collateral_asset),
+				Error::<T>::InvalidLoanParameters
+			);
+
+			account.primary_collateral_asset = primary_collateral_asset;
+
+			account.try_adding_collateral_from_free_balance(&borrower_id, &extra_collateral)?;
+
+			ensure!(
+				account.derive_ltv()? <= config.ltv_thresholds.target,
+				Error::<T>::InsufficientCollateral
+			);
+
+			GeneralLendingPools::<T>::try_mutate(asset, |pool| {
+				let pool = pool.as_mut().ok_or(Error::<T>::PoolDoesNotExist)?;
+
+				pool.provide_funds_for_loan(amount_to_borrow)?;
```

Comment:
From an API / correctness point of view, it might be nice to explicitly link loan creation with the pool to satisfy ourselves that taking a loan requires interaction with a lending pool. (ie. a method on GeneralLoan that takes a pool as argument). 

And if we structure it as eg. `GeneralLoan::new(..).expand(pool, loan_amount)` then we can share much more code between `new_loan` and `expand_loan`.

Reply:
`expand_loan_inner` is a little awkward in that expanding a loan requires removing it from the map so that it can be inserted again... it seems like it could take just an id, and then expand or create as necessary.

In fact, what's the reason for differentiating between expand and create? Aren't they essentially the same operations? The only difference seems to be that expand can fail if you accidentally provide the wrong loan id...

Reply:
I tried moving loan creation into`expand_loan` and ended up with a similar borrow checked problem: you end up having to hold a mut reference (in case the loan already existed), while also modifying the loan_account. There are ways to address this of course (e.g. inlining `try_adding_collateral_from_free_balance`, turning it into a free standing function, or reading the loan multiple times), but they are all felt like compromises, and not an improvement over what I had initially, so I think in the end I prefer to leave the code as is.